app/src/main/java/com/antonioleiva/bandhookkotlin/repository/ArtistRepositoryImp.kt

```diff
@@ -16,12 +16,36 @@
 
 package com.antonioleiva.bandhookkotlin.repository
 
+import com.antonioleiva.bandhookkotlin.Result
+import com.antonioleiva.bandhookkotlin.ResultT
 import com.antonioleiva.bandhookkotlin.domain.entity.Artist
+import com.antonioleiva.bandhookkotlin.domain.entity.BizException.ArtistNotFound
 import com.antonioleiva.bandhookkotlin.domain.repository.ArtistRepository
+import com.antonioleiva.bandhookkotlin.recoverWith
 import com.antonioleiva.bandhookkotlin.repository.dataset.ArtistDataSet
+import org.funktionale.collections.tail
 
 class ArtistRepositoryImp(val artistDataSets: List<ArtistDataSet>) : ArtistRepository {
 
+    /** Trampolined (async) recursion until all DataSets are exhausted or a right result is found */
+    fun <E, A> recurseDataSets(currentDataSets: List<ArtistDataSet>, acc: Result<E, A>, f:
```

Comment:
We should be able to generalize this for all datasets and place this function in the top level Repository interface